driver/src/price_estimation/threaded_price_source.rs

```diff
@@ -0,0 +1,134 @@
+    // Allows the thread to notice when the owning struct is dropped.
+    _channel_to_thread: Sender<()>,
+}
+
+impl ThreadedPriceSource {
+    /// All token prices will be updated every `update_interval`. Prices for
+    /// other tokens will not be returned in `get_prices`.
+    ///
+    /// The join handle represents the background thread. It can be used to
+    /// verify that it exits when the created struct is dropped.
+    #[allow(dead_code)]
+    pub fn new<T: 'static + PriceSource + Send>(
+        tokens: Vec<Token>,
+        price_source: T,
+        update_interval: Duration,
+    ) -> (Self, JoinHandle<()>) {
+        let price_map = Arc::new(Mutex::new(HashMap::new()));
+        let (sender, receiver) = mpsc::channel::<()>();
+
+        let price_map_ = price_map.clone();
+        // vk: Clippy notes that the following loop and match can be written as
+        // a `while let` loop but I find it clearer to make all cases explicit.
+        #[allow(clippy::while_let_loop)]
+        let join_handle = thread::spawn(move || loop {
+            match receiver.recv_timeout(update_interval) {
+                Ok(()) | Err(RecvTimeoutError::Timeout) => {
+                    // Make sure we don't hold the mutex while this blocking call happens.
+                    match price_source.get_prices(&tokens) {
+                        Ok(prices) => price_map_.lock().unwrap().extend(prices),
+                        Err(err) => log::warn!("price_source::get_prices failed: {}", err),
+                    }
+                }
+                // The owning struct has been dropped.
+                Err(RecvTimeoutError::Disconnected) => break,
+            }
+        });
+
+        // Update the prices immediately.
+        sender.send(()).unwrap();
+        (
+            Self {
+                price_map,
+                _channel_to_thread: sender,
+            },
+            join_handle,
+        )
+    }
+}
+
+impl PriceSource for ThreadedPriceSource {
+    /// Non blocking.
+    /// Infallible.
+    fn get_prices(&self, _tokens: &[Token]) -> Result<HashMap<TokenId, u128>> {
+        Ok(self.price_map.lock().unwrap().clone())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::price_source::MockPriceSource;
+    use super::*;
+    use std::sync::atomic;
+
+    const ORDERING: atomic::Ordering = atomic::Ordering::SeqCst;
+
+    lazy_static::lazy_static! {
+        static ref TOKENS: [Token; 3] = [
+            Token::new(0, "0", 0),
+            Token::new(1, "1", 1),
+            Token::new(2, "2", 2),
+        ];
+    }
+
+    /// Drops tps and joins the handle which ensures that the thread exits and
+    /// hasn't panicked. This is important because the mockall expectation
+    /// panics happen in the thread which would otherwise let the tests pass
+    /// when an expectation is violated.
+    fn join(tps: ThreadedPriceSource, handle: JoinHandle<()>) {
+        std::mem::drop(tps);
+        handle.join().unwrap();
```

Comment:
It's not great that we'd wait forever if the thread doesn't finish (specfically for CI where I don't believe we have timeouts). 
Would it be a lot of effort to signal back to the owning struct (either via a channel or a boolean) that we have finished running?

Reply:
I can hardcode a timeout by doing the joining in another thread and having the main thread panic if that thread does not exist within some time. That does have the brittle `sleep` again but it is better than hanging forever.